wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -576,20 +580,36 @@ private String sessionUserNameOr(String nameIfNoUser) {
     return nameIfNoUser;
   }
 
-  private class CleanupRuntimeOnAbnormalRuntimeStop implements EventSubscriber<RuntimeStatusEvent> {
+  private class AbnormalRuntimeStopListener implements EventSubscriber<RuntimeStatusEvent> {
     @Override
     public void onEvent(RuntimeStatusEvent event) {
       if (event.isFailed()) {
         RuntimeState state = runtimes.remove(event.getIdentity().getWorkspaceId());
         if (state != null) {
+          String workspaceId = state.runtime.getContext().getIdentity().getWorkspaceId();
           publishWorkspaceStatusEvent(
-              state.runtime.getContext().getIdentity().getWorkspaceId(),
+              workspaceId,
               STOPPED,
               RUNNING,
               "Error occurs on workspace runtime stop. Error: " + event.getError());
+          setAbnormalStopAttributes(workspaceId, event.getError());
         }
       }
     }
+
+    private void setAbnormalStopAttributes(String workspaceId, String error) {
```

Comment:
Is this change covered by unit tests?

Reply:
Hell..i have no idea how to test this inner class yet..

Reply:
1. Call WorkspaceRunitmes#init to subscribe abnormal stop listener to the EventService
1. Call EventService#publish to fire RuntimeStatusEvent about failed runtime
1. Verify event workspace event publishing, an update of workspace attributes.

Please, do not hesitate to comment back if I missed something